app/models/dataset_record/darwin_core/taxon.rb

```diff
@@ -225,13 +254,33 @@ def import(dwc_data_attributes = {})
           if parent.is_a?(Combination)
             parent_elements = parent.combination_taxon_names.index_by { |protonym| protonym.rank }
 
-          else  # parent is a protonym, so we need to get all parent elements up to genus
-            parent_elements = parent.self_and_ancestors.order('taxon_name_hierarchies.generations DESC').to_a
-                                    .take_while{ |p| ORIGINAL_COMBINATION_RANKS.has_key?(p&.rank&.to_sym) }
-                                    .index_by{ |protonym| protonym.rank}
+          else
+            # parent is a protonym, look at parents in checklist to build combination
+
+            parents = [get_parent]
+
+            while (next_parent = find_by_taxonID(parents[-1].metadata['parent']))
+              parents << next_parent
+            end
+
+            # convert DatasetRecords into hash of rank, protonym pairs
+            parent_elements = parents.to_h do |p|
+              [
+                # Key is rank (as set in checklist file)
+                DatasetRecordField.where(dataset_record_id: p)
+                                  .at(get_field_mapping(:taxonRank))
+                                  .pick(:value)
```

Comment:
From here on safe navigation should be used, in case `get_field_mapping(:taxonRank)` and/or `.pick(:value)` return `nil` (former perhaps is prevented from happening somewhere else, but latter can definitively happen).